### ISSUE TITLE
DataLoader returning non-CUDA tensors

### ISSUE THREAD

**Issue by abhaikollara** [pytorch/pytorch#5398]
I created a simple dataset that has two CUDA tensors, one of shape `(1000, 5)` and the other `(1000,)` but when I load them using a `DataLoader` only the first one is returned as CUDA tensor. I think this may have something to do with the default `collate_fn` of the DataLoader. A simple hack is to reshape the 1D tensor into `(1000, 1)` while feeding to DataLoader. But loss functions like `CrossEntropyLoss` and `NLLLoss` require 1D inputs which means they have to reshaped back into `(1000, )`. Are there any proper fixes to this ?

```python
import torch
from torch.utils.data import Dataset, DataLoader
import numpy as np

x = torch.cuda.FloatTensor(np.random.randn(1000, 5))
y = torch.cuda.LongTensor(np.random.randint(0,10, size=(1000, )))

class data(Dataset):
    def __init__(self, inputs, targets):
        self.x = inputs
        self.y = targets

    def __len__(self):
        return self.x.size()[0]

    def __getitem__(self, idx):
         return (self.x[idx], self.y[idx])

d = data(x, y)
dl = DataLoader(d, batch_size=5)

for item in dl:
    print(item)
    break
```
Output
```
[
 0.9943 -0.6458 -0.7077 -0.9228 -1.1230
-0.1050 -0.1452  0.0680 -0.5407 -0.5932
 2.7039  0.0426  0.0692  1.3255  0.0133
-0.3883  0.5905  1.4192 -0.7205 -1.5623
 1.8158  0.5436  0.8788  2.3459  0.9836
[torch.cuda.FloatTensor of size 5x5 (GPU 0)]
,
 6
 7
 2
 2
 7
[torch.LongTensor of size 5]
]
```

- OS: Ubuntu 17.10
- PyTorch version: 0.3.0.post4
- How you installed PyTorch (conda, pip, source): pip
- Python version: Python 3.6.3
- CUDA/cuDNN version: 8.0.61

**Comment by sampathweb** [comment]
@abhaikollara - Interesting.  Possibly because when you do `self.y[idx]`, you get a Scalar value of type int and not an torch.cuda tensor.
```
>>> print(type(self.y[idx]), type(self.x[idx]))`
<class 'int'> <class 'torch.cuda.FloatTensor'>
```
Alternatively, during the training loop, you might want to put X and y in cuda while wrapping it in a variable like most pyTorch code examples.  Hope this is useful

**Comment by abhaikollara** [comment]
@sampathweb Yeah, the issue is that the default collate function of the DataLoader selects individual scalar values and combines them.

And as far as the wrapping it in cuda during training loop, I'm not sure how efficient that is compared to a single move of the entire data to GPU

**Comment by SsnL** [comment]
If you `pin_memory`, it will be pretty fast. This will be fixed with what we have recently done with scalar and merging Variable and Tensor.

**Comment by abhaikollara** [comment]
@SsnL Could you clarify the usage of `pin_memory` I'm not sure I understand how it works.

**Comment by SsnL** [comment]
@abhaikollara It reserves an area in CPU memory so pulling tensors in that area to GPU is fast. See http://pytorch.org/docs/master/notes/cuda.html#use-pinned-memory-buffers

**Comment by apaszke** [comment]
This is already fixed in master. Thanks for the report!
